src/System.Memory/src/System/Span.cs

```diff
@@ -323,37 +322,64 @@ public void CopyTo(Span<T> destination)
         /// <param name="destination">The span to copy items into.</param>
         public bool TryCopyTo(Span<T> destination)
         {
-            if ((uint)_length > (uint)destination._length)
+            int length = _length;
+            int destLength = destination._length;
+
+            if ((uint)length > (uint)destLength)
                 return false;
 
-            // TODO: This is a tide-over implementation as we plan to add a overlap-safe cpblk-based api to Unsafe. (https://github.com/dotnet/corefx/issues/13427)
             unsafe
             {
                 ref T src = ref DangerousGetPinnableReference();
                 ref T dst = ref destination.DangerousGetPinnableReference();
                 IntPtr srcMinusDst = Unsafe.ByteOffset<T>(ref dst, ref src);
-                int length = _length;
+                IntPtr dstMinusSrc = Unsafe.ByteOffset<T>(ref src, ref dst);
 
                 bool srcGreaterThanDst = (sizeof(IntPtr) == sizeof(int)) ? srcMinusDst.ToInt32() >= 0 : srcMinusDst.ToInt64() >= 0;
                 if (srcGreaterThanDst)
                 {
-                    // Source address greater than or equal to destination address. Can do normal copy.
-                    for (int i = 0; i < length; i++)
+                    IntPtr tailDiff = Unsafe.ByteOffset<T>(ref Unsafe.Add<T>(ref dst, destLength), ref src);
+                    bool isOverlapped = (sizeof(IntPtr) == sizeof(int)) ? tailDiff.ToInt32() < 0 : tailDiff.ToInt64() < 0;
+                    if (!isOverlapped)
                     {
-                        Unsafe.Add<T>(ref dst, i) = Unsafe.Add<T>(ref src, i);
+                        Unsafe.CopyBlock<T>(ref dst, ref src, (uint)length);
                     }
+                    else
+                    {
+                        // Source address greater than or equal to destination address. Can do normal copy.
+                        for (int i = 0; i < length; i++)
+                        {
+                            Unsafe.Add<T>(ref dst, i) = Unsafe.Add<T>(ref src, i);
```

Comment:
Either a different commit or a different PR, but I think we should try to make the gap between overlapping and non-overlapping as small as possible. Otherwise it creates annoying performance traps.